File: src/Bindings.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "gqlmapi-rs/include/Bindings.h"
#include "gqlmapi-rs/src/bindings.rs.h"

#include "gqlmapi-rs/include/ResponseTypes.h"

#include "MAPIGraphQL.h"

#include <iostream>
#include <memory>
#include <map>
#include <optional>
#include <queue>
#include <thread>

using namespace graphql;

class Subscription : public std::enable_shared_from_this<Subscription>
{
public:
	explicit Subscription(const std::shared_ptr<service::Request> &service,
						  rust::Box<NextContext> nextContext,
						  Bindings::NextCallback nextCallback,
						  rust::Box<CompleteContext> completeContext,
						  Bindings::CompleteCallback completeCallback) noexcept;
	~Subscription();

	void Subscribe(service::SubscriptionKey key) noexcept;
	void Unsubscribe();
	void Deliver(response::AwaitableValue &&payload);
	void Deliver(response::Value &&document);
	void Complete();

private:
	std::weak_ptr<service::Request> _weakService;
	rust::Box<NextContext> _nextContext;
	Bindings::NextCallback _nextCallback;
	rust::Box<CompleteContext> _completeContext;
	Bindings::CompleteCallback _completeCallback;
	std::optional<service::SubscriptionKey> _key = std::nullopt;
	bool _registered = false;
};

Subscription::Subscription(const std::shared_ptr<service::Request> &service,
						   rust::Box<NextContext> nextContext,
						   Bindings::NextCallback nextCallback,
						   rust::Box<CompleteContext> completeContext,
						   Bindings::CompleteCallback completeCallback) noexcept
	: _weakService{service}, _nextContext{std::move(nextContext)}, _nextCallback{std::move(nextCallback)}, _completeContext{std::move(completeContext)}, _completeCallback{std::move(completeCallback)}
{
}

Subscription::~Subscription()
{
	Unsubscribe();
}

void Subscription::Subscribe(service::SubscriptionKey key) noexcept
{
	_registered = true;
	_key = std::make_optional(key);
}

void Subscription::Unsubscribe()
{
	if (!_registered)
	{
		return;
	}

	_registered = false;

	auto deferUnsubscribe = std::move(_key);
	auto service = _weakService.lock();

	if (deferUnsubscribe && service)
	{
		service->unsubscribe({*deferUnsubscribe}).get();
		Complete();
	}
}

void Subscription::Deliver(response::AwaitableValue &&payload)
{
	response::Value document{response::Type::Map};

	try
	{
		document = payload.get();
	}
	catch (service::schema_exception &scx)
	{
		document.reserve(2);
		document.emplace_back(std::string{service::strData}, {});
		document.emplace_back(std::string{service::strErrors}, scx.getErrors());
	}
	catch (const std::exception &ex)
	{
		std::ostringstream oss;

		oss << "Caught exception delivering subscription payload: " << ex.what();
		document.reserve(2);
		document.emplace_back(std::string{service::strData}, {});
		document.emplace_back(std::string{service::strErrors}, response::Value{oss.str()});
	}

	Deliver(std::move(document));
}

void Subscription::Deliver(response::Value &&document)
{
	_nextContext = _nextCallback(std::move(_nextContext), toJSON(std::move(document)));
}

void Subscription::Complete()
{
	_completeCallback(std::move(_completeContext));
}

class RegisteredSubscription
{
public:
	explicit RegisteredSubscription(const std::shared_ptr<service::Request> &service,
									peg::ast &ast,
									std::string_view operationName,
									response::Value &&variables,
									rust::Box<NextContext> nextContext,
									Bindings::NextCallback nextCallback,
									rust::Box<CompleteContext> completeContext,
									Bindings::CompleteCallback completeCallback);

	void Unsubscribe() noexcept;

private:
	std::shared_ptr<Subscription> _subscription;
};

RegisteredSubscription::RegisteredSubscription(const std::shared_ptr<service::Request> &service,
											   peg::ast &ast,
											   std::string_view operationName,
											   response::Value &&variables,
											   rust::Box<NextContext> nextContext,
											   Bindings::NextCallback nextCallback,
											   rust::Box<CompleteContext> completeContext,
											   Bindings::CompleteCallback completeCallback)
	: _subscription{std::make_shared<Subscription>(service,
												   std::move(nextContext),
												   std::move(nextCallback),
												   std::move(completeContext),
												   std::move(completeCallback))}
{
	if (service->findOperationDefinition(ast, operationName).first == service::strSubscription)
	{
		_subscription->Subscribe(
			service->subscribe({[weakSubscription = std::weak_ptr{_subscription}](response::Value payload)
								{
									auto subscription = weakSubscription.lock();

									if (subscription)
									{
										subscription->Deliver(std::move(payload));
									}
								},
								peg::ast{ast},
								std::string{operationName},
								std::move(variables)})
				.get());
	}
	else
	{
		_subscription->Deliver(service->resolve({ast,
												 operationName,
												 std::move(variables)}));
		_subscription->Complete();
	}
}

void RegisteredSubscription::Unsubscribe() noexcept
{
	const auto subscription = std::move(_subscription);

	if (subscription)
	{
		subscription->Unsubscribe();
	}
}

class Bindings::impl
{
public:
	impl() = default;
	~impl() = default;

	void startService(bool useDefaultProfile) noexcept;
	void stopService();

	std::int32_t parseQuery(std::string_view query);
	void discardQuery(std::int32_t queryId) noexcept;

	std::int32_t subscribe(std::int32_t queryId,
						   std::string_view operationName,
						   std::string_view variables,
						   rust::Box<NextContext> nextContext,
						   NextCallback nextCallback,
						   rust::Box<CompleteContext> completeContext,
						   CompleteCallback completeCallback);
	void unsubscribe(std::int32_t subscriptionId);

private:
	std::shared_ptr<service::Request> service;
	std::map<std::int32_t, peg::ast> queryMap;
	std::map<std::int32_t, std::unique_ptr<RegisteredSubscription>> subscriptionMap;
};

void Bindings::impl::startService(bool useDefaultProfile) noexcept
{
	service = mapi::GetService(useDefaultProfile);
}

void Bindings::impl::stopService()
{
	if (service)
	{
		for (const auto &entry : subscriptionMap)
		{
			entry.second->Unsubscribe();
		}

		subscriptionMap.clear();
		queryMap.clear();
		service.reset();
	}
}

std::int32_t Bindings::impl::parseQuery(std::string_view query)
{
	const std::int32_t queryId = (queryMap.empty() ? 1 : queryMap.crbegin()->first + 1);

	queryMap[queryId] = peg::parseString(query);
	return queryId;
}

void Bindings::impl::discardQuery(std::int32_t queryId) noexcept
{
	queryMap.erase(queryId);
}

std::int32_t Bindings::impl::subscribe(std::int32_t queryId,
									   std::string_view operationName,
									   std::string_view variables,
									   rust::Box<NextContext> nextContext,
									   NextCallback nextCallback,
									   rust::Box<CompleteContext> completeContext,
									   CompleteCallback completeCallback)
{
	const auto itrQuery = queryMap.find(queryId);

	if (itrQuery == queryMap.cend())
	{
		throw std::runtime_error("Unknown queryId");
	}

	auto &ast = itrQuery->second;
	auto parsedVariables = (variables.empty() ? response::Value(response::Type::Map) : parseJSON(variables));

	if (parsedVariables.type() != response::Type::Map)
	{
		throw std::runtime_error("Invalid variables object");
	}

	if (!service)
	{
		throw std::runtime_error("Did not call startService");
	}

	const std::int32_t subscriptionId = (subscriptionMap.empty() ? 1 : subscriptionMap.crbegin()->first + 1);

	subscriptionMap[subscriptionId] = std::make_unique<RegisteredSubscription>(service,
																			   ast,
																			   operationName,
																			   std::move(parsedVariables),
																			   std::move(nextContext),
																			   std::move(nextCallback),
																			   std::move(completeContext),
																			   std::move(completeCallback));

	return subscriptionId;
}

void Bindings::impl::unsubscribe(std::int32_t subscriptionId)
{
	auto itr = subscriptionMap.find(subscriptionId);

	if (itr != subscriptionMap.end())
	{
		itr->second->Unsubscribe();
		subscriptionMap.erase(itr);
	}
}

Bindings::Bindings() noexcept
	: m_pimpl{std::make_unique<impl>()}
{
}

Bindings::~Bindings()
{
}

void Bindings::startService(bool useDefaultProfile) const noexcept
{
	m_pimpl->startService(useDefaultProfile);
}

void Bindings::stopService() const
{
	m_pimpl->stopService();
}

std::int32_t Bindings::parseQuery(rust::Str query) const
{
	return m_pimpl->parseQuery(std::string_view{query.data(), query.size()});
}

void Bindings::discardQuery(std::int32_t queryId) const noexcept
{
	m_pimpl->discardQuery(queryId);
}

std::int32_t Bindings::subscribe(std::int32_t queryId,
								 rust::Str operationName,
								 rust::Str variables,
								 rust::Box<NextContext> nextContext,
								 NextCallback nextCallback,
								 rust::Box<CompleteContext> completeContext,
								 CompleteCallback completeCallback) const
{
	return m_pimpl->subscribe(queryId,
							  std::string_view{operationName.data(), operationName.size()},
							  std::string_view{variables.data(), variables.size()},
							  std::move(nextContext),
							  std::move(nextCallback),
							  std::move(completeContext),
							  std::move(completeCallback));
}

void Bindings::unsubscribe(std::int32_t subscriptionId) const
{
	m_pimpl->unsubscribe(subscriptionId);
}

std::unique_ptr<Bindings> make_bindings() noexcept
{
	return std::make_unique<Bindings>();
}

ResponseValue::ResponseValue(ResponseValueType type)
	: m_impl{type}
{
}

ResponseValue::ResponseValue(response::Value &&other)
	: m_impl{std::move(other)}
{
}

ResponseValueType ResponseValue::getType() const noexcept
{
	return m_impl.type();
}

ResponseValue &ResponseValue::fromJson() noexcept
{
	m_impl = m_impl.from_json();
	return *this;
}

void ResponseValue::reserve(size_t additional)
{
	m_impl.reserve(m_impl.size() + additional);
}

bool ResponseValue::pushMapEntry(rust::Str name, std::unique_ptr<ResponseValue> value)
{
	return m_impl.emplace_back(std::string{name.data(), name.size()}, std::move(value->m_impl));
}

void ResponseValue::pushListEntry(std::unique_ptr<ResponseValue> value)
{
	m_impl.emplace_back(std::move(value->m_impl));
}

void ResponseValue::setString(rust::Str value)
{
	m_impl.set<response::StringType>(response::StringType{value.data(), value.size()});
}

void ResponseValue::setBool(bool value)
{
	m_impl.set<response::BooleanType>(value);
}

void ResponseValue::setInt(std::int64_t value)
{
	m_impl.set<response::IntType>(value);
}

void ResponseValue::setFloat(double value)
{
	m_impl.set<response::FloatType>(value);
}

std::unique_ptr<std::vector<ResponseMapEntry>> ResponseValue::releaseMap()
{
	auto data = m_impl.release<response::MapType>();
	auto entries = std::make_unique<std::vector<ResponseMapEntry>>();

	if (!data.empty())
	{
		entries->reserve(data.size());
		std::transform(data.begin(), data.end(), std::back_inserter(*entries), [](auto &member) noexcept
					   { return ResponseMapEntry{
							 std::make_unique<std::string>(std::move(member.first)),
							 std::make_unique<ResponseValue>(std::move(member.second))}; });
	}

	return entries;
}

std::unique_ptr<std::vector<ResponseValue>> ResponseValue::releaseList()
{
	auto data = m_impl.release<response::ListType>();
	auto entries = std::make_unique<std::vector<ResponseValue>>();

	if (!data.empty())
	{
		entries->reserve(data.size());
		std::transform(data.begin(), data.end(), std::back_inserter(*entries), [](auto &member) noexcept
					   { return ResponseValue{std::move(member)}; });
	}

	return entries;
}

std::unique_ptr<std::string> ResponseValue::releaseString()
{
	return std::make_unique<std::string>(m_impl.release<response::StringType>());
}

bool ResponseValue::getBool() const
{
	return m_impl.get<response::BooleanType>();
}

std::int64_t ResponseValue::getInt() const
{
	return m_impl.get<response::IntType>();
}

double ResponseValue::getFloat() const
{
	return m_impl.get<response::FloatType>();
}

std::unique_ptr<ResponseValue> ResponseValue::releaseScalar()
{
	return std::make_unique<ResponseValue>(m_impl.release<response::ScalarType>());
}

graphql::response::Value ResponseValue::releaseValue() noexcept
{
	return {std::move(m_impl)};
}

std::unique_ptr<ResponseValue> makeResponseValue(ResponseValueType type) noexcept
{
	return std::make_unique<ResponseValue>(type);
}

rust::String toJSON(graphql::response::Value &&document)
{
	ResponseValue payload{std::move(document)};
	return from_value(payload)->to_json();
}

graphql::response::Value parseJSON(std::string_view document)
{
	return parse_json(rust::Str{document.data(), document.size()})->into_value()->releaseValue();
}

File: src/bindings.rs
use std::{pin::Pin, str::FromStr};

use serde_json::Value;

#[cxx::bridge]
pub mod ffi {
    extern "Rust" {
        type NextContext;
        type CompleteContext;
    }

    enum ResponseValueType {
        Map,
        List,
        String,
        Null,
        Boolean,
        Int,
        Float,
        EnumValue,
        ID,
        Scalar,
    }

    struct ResponseMapEntry {
        name: UniquePtr<CxxString>,
        value: UniquePtr<ResponseValue>,
    }

    unsafe extern "C++" {
        include!("gqlmapi-rs/include/ResponseTypes.h");
        type ResponseValueType;

        type ResponseValue;

        #[cxx_name = "makeResponseValue"]
        fn make_response_value(value_type: ResponseValueType) -> UniquePtr<ResponseValue>;

        #[cxx_name = "getType"]
        fn get_type(self: &ResponseValue) -> ResponseValueType;
        #[cxx_name = "fromJson"]
        fn from_json(self: Pin<&mut ResponseValue>) -> Pin<&mut ResponseValue>;
        fn reserve(self: Pin<&mut ResponseValue>, additional: usize) -> Result<()>;
        #[cxx_name = "pushMapEntry"]
        fn push_map_entry(
            self: Pin<&mut ResponseValue>,
            name: &str,
            value: UniquePtr<ResponseValue>,
        ) -> Result<bool>;
        #[cxx_name = "pushListEntry"]
        fn push_list_entry(
            self: Pin<&mut ResponseValue>,
            value: UniquePtr<ResponseValue>,
        ) -> Result<()>;
        #[cxx_name = "setString"]
        fn set_string(self: Pin<&mut ResponseValue>, value: &str) -> Result<()>;
        #[cxx_name = "setBool"]
        fn set_bool(self: Pin<&mut ResponseValue>, value: bool) -> Result<()>;
        #[cxx_name = "setInt"]
        fn set_int(self: Pin<&mut ResponseValue>, value: i64) -> Result<()>;
        #[cxx_name = "setFloat"]
        fn set_float(self: Pin<&mut ResponseValue>, value: f64) -> Result<()>;
        #[cxx_name = "releaseMap"]
        fn release_map(
            self: Pin<&mut ResponseValue>,
        ) -> Result<UniquePtr<CxxVector<ResponseMapEntry>>>;
        #[cxx_name = "releaseList"]
        fn release_list(
            self: Pin<&mut ResponseValue>,
        ) -> Result<UniquePtr<CxxVector<ResponseValue>>>;
        #[cxx_name = "releaseString"]
        fn release_string(self: Pin<&mut ResponseValue>) -> Result<UniquePtr<CxxString>>;
        #[cxx_name = "getBool"]
        fn get_bool(self: &ResponseValue) -> Result<bool>;
        #[cxx_name = "getInt"]
        fn get_int(self: &ResponseValue) -> Result<i64>;
        #[cxx_name = "getFloat"]
        fn get_float(self: &ResponseValue) -> Result<f64>;
        #[cxx_name = "releaseScalar"]
        fn release_scalar(self: Pin<&mut ResponseValue>) -> Result<UniquePtr<ResponseValue>>;
    }

    extern "Rust" {
        type JsonValue;

        fn parse_json(json: &str) -> Result<Box<JsonValue>>;
        fn to_json(&mut self) -> Result<String>;

        fn from_value(value: Pin<&mut ResponseValue>) -> Result<Box<JsonValue>>;
        fn into_value(&mut self) -> Result<UniquePtr<ResponseValue>>;
    }

    unsafe extern "C++" {
        include!("gqlmapi-rs/include/Bindings.h");

        type Bindings;

        fn make_bindings() -> UniquePtr<Bindings>;

        fn startService(&self, useDefaultProfile: bool);
        fn stopService(&self);

        fn parseQuery(&self, query: &str) -> Result<i32>;
        fn discardQuery(&self, queryId: i32);

        #[allow(clippy::too_many_arguments)]
        fn subscribe(
            &self,
            queryId: i32,
            operationName: &str,
            variables: &str,
            nextContext: Box<NextContext>,
            nextCallback: fn(Box<NextContext>, String) -> Box<NextContext>,
            completeContext: Box<CompleteContext>,
            completeCallback: fn(Box<CompleteContext>),
        ) -> Result<i32>;
        fn unsubscribe(&self, subscriptionId: i32);
    }
}

pub struct NextContext {
    pub callback: Box<dyn FnMut(String)>,
    pub thread_id: u32,
}

pub struct CompleteContext {
    pub callback: Box<dyn FnOnce()>,
    pub thread_id: u32,
}

struct JsonValue(Option<Value>);

fn parse_json(json: &str) -> Result<Box<JsonValue>, String> {
    JsonValue::new(json).map_err(|err| err.to_string())
}

fn from_value(value: Pin<&mut ffi::ResponseValue>) -> Result<Box<JsonValue>, String> {
    JsonValue::try_from(value).map(Box::new)
}

impl JsonValue {
    fn new(json: &str) -> Result<Box<Self>, serde_json::Error> {
        let value = Value::from_str(json)?;
        Ok(Box::new(Self(Some(value))))
    }

    #[allow(clippy::wrong_self_convention)]
    fn to_json(&mut self) -> Result<String, serde_json::Error> {
        let value = self.0.take().unwrap_or(Value::Null);
        serde_json::to_string(&value)
    }

    #[allow(clippy::wrong_self_convention)]
    fn into_value(&mut self) -> Result<cxx::UniquePtr<ffi::ResponseValue>, String> {
        let value = JsonValue(self.0.take());
        value.try_into()
    }
}

impl TryInto<cxx::UniquePtr<ffi::ResponseValue>> for JsonValue {
    type Error = String;

    fn try_into(self) -> Result<cxx::UniquePtr<ffi::ResponseValue>, String> {
        Ok(match self.0 {
            None | Some(Value::Null) => {
                let result = ffi::make_response_value(ffi::ResponseValueType::Null);
                result
                    .as_ref()
                    .ok_or("Failed to allocate Null ResponseValue".to_owned())?;
                result
            }
            Some(Value::Bool(value)) => {
                let mut result = ffi::make_response_value(ffi::ResponseValueType::Boolean);
                result
                    .as_mut()
                    .ok_or("Failed to allocate Bool ResponseValue".to_owned())?
                    .set_bool(value)
                    .map_err(|err| format!("Failed to set Bool: {err}"))?;
                result
            }
            Some(Value::Number(value)) => {
                if value.is_i64() {
                    let mut result = ffi::make_response_value(ffi::ResponseValueType::Int);
                    result
                        .as_mut()
                        .ok_or("Failed to allocate Int ResponseValue".to_owned())?
                        .set_int(value.as_i64().ok_or("Int value out of bounds".to_owned())?)
                        .map_err(|err| format!("Failed to set Int: {err}"))?;
                    result
                } else if value.is_f64() {
                    let mut result = ffi::make_response_value(ffi::ResponseValueType::Float);
                    result
                        .as_mut()
                        .ok_or("Failed to allocate Float ResponseValue".to_owned())?
                        .set_float(
                            value
                                .as_f64()
                                .ok_or("Float value out of bounds".to_owned())?,
                        )
                        .map_err(|err| format!("Failed to set Float: {err}"))?;
                    result
                } else {
                    Err(format!("Unrecognized Number: {value}"))?
                }
            }
            Some(Value::String(value)) => {
                let mut result = ffi::make_response_value(ffi::ResponseValueType::String);
                result
                    .as_mut()
                    .ok_or("Failed to allocate String ResponseValue".to_owned())?
                    .from_json()
                    .set_string(&value)
                    .map_err(|err| format!("Failed to set String: {err}"))?;
                result
            }
            Some(Value::Array(value)) => {
                let mut result = ffi::make_response_value(ffi::ResponseValueType::List);
                let mut pinned = result
                    .as_mut()
                    .ok_or("Failed to allocate List ResponseValue".to_owned())?;

                if !value.is_empty() {
                    pinned.as_mut().reserve(value.len()).map_err(|err| {
                        format!(
                            "Failed to reserve List with capacity {}: {err}",
                            value.len()
                        )
                    })?;
                    for (i, value) in value.into_iter().enumerate() {
                        let value = JsonValue(Some(value)).try_into()?;
                        pinned
                            .as_mut()
                            .push_list_entry(value)
                            .map_err(|err| format!("Failed to push entry {i} into List: {err}"))?;
                    }
                }
                result
            }
            Some(Value::Object(value)) => {
                let mut result = ffi::make_response_value(ffi::ResponseValueType::Map);
                let mut pinned = result
                    .as_mut()
                    .ok_or("Failed to allocate Map ResponseValue".to_owned())?;

                if !value.is_empty() {
                    pinned.as_mut().reserve(value.len()).map_err(|err| {
                        format!("Failed to reserve Map with capacity {}: {err}", value.len())
                    })?;
                    for (name, value) in value.into_iter() {
                        let value = JsonValue(Some(value)).try_into()?;
                        pinned
                            .as_mut()
                            .push_map_entry(&name, value)
                            .map_err(|err| {
                                format!("Failed to push entry \"{name}\" into Map: {err}")
                            })?;
                    }
                }
                result
            }
        })
    }
}

impl TryFrom<Pin<&mut ffi::ResponseValue>> for JsonValue {
    type Error = String;

    fn try_from(mut value: Pin<&mut ffi::ResponseValue>) -> Result<Self, String> {
        Ok(Self(Some(match value.as_mut().get_type() {
            ffi::ResponseValueType::Map => {
                let mut members = value
                    .as_mut()
                    .release_map()
                    .map_err(|err| format!("Failed to release Map entries: {err}"))?;
                let members = members
                    .as_mut()
                    .ok_or("Map ResponseValue returned a null vector".to_owned())?;
                let mut map = serde_json::Map::new();
                for ffi::ResponseMapEntry { name, value } in members.as_mut_slice() {
                    if let (Some(name), Some(value)) = (name.as_ref(), value.as_mut()) {
                        if let (Ok(name), Ok(JsonValue(Some(value)))) =
                            (name.to_str(), value.try_into())
                        {
                            map.insert(name.to_owned(), value);
                        }
                    }
                }
                Value::Object(map)
            }
            ffi::ResponseValueType::List => {
                let mut members = value
                    .as_mut()
                    .release_list()
                    .map_err(|err| format!("Failed to release List entries: {err}"))?;
                let members = members
                    .as_mut()
                    .ok_or("List ResponseValue returned a null vector".to_owned())?;
                let mut list = Vec::new();
                for value in members.iter_mut() {
                    if let Ok(JsonValue(Some(value))) = value.try_into() {
                        list.push(value);
                    }
                }
                Value::Array(list)
            }
            ffi::ResponseValueType::String
            | ffi::ResponseValueType::EnumValue
            | ffi::ResponseValueType::ID => {
                if let Ok(value) = value
                    .as_mut()
                    .release_string()
                    .map_err(|err| format!("Failed to release String: {err}"))?
                    .as_mut()
                    .ok_or("String ResponseValue returned a null value".to_owned())?
                    .to_str()
                {
                    Value::String(value.to_owned())
                } else {
                    Value::Null
                }
            }
            ffi::ResponseValueType::Null => Value::Null,
            ffi::ResponseValueType::Boolean => Value::Bool(
                value
                    .as_mut()
                    .get_bool()
                    .map_err(|err| format!("Failed to get Boolean: {err}"))?,
            ),
            ffi::ResponseValueType::Int => {
                let value = value
                    .as_mut()
                    .get_int()
                    .map_err(|err| format!("Failed to get Int: {err}"))?;
                serde_json::json!(value)
            }
            ffi::ResponseValueType::Float => {
                let value = value
                    .as_mut()
                    .get_float()
                    .map_err(|err| format!("Failed to get Float: {err}"))?;
                serde_json::json!(value)
            }
            ffi::ResponseValueType::Scalar => {
                let mut value = value
                    .as_mut()
                    .release_scalar()
                    .map_err(|err| format!("Failed to release Scalar: {err}"))?;
                let value = value
                    .as_mut()
                    .ok_or("Scalar ResponseValue returned a null value".to_owned())?;
                if let Ok(JsonValue(Some(value))) = value.try_into() {
                    value
                } else {
                    Value::Null
                }
            }
            _ => unreachable!(),
        })))
    }
}
